libpromises/rlist.c

```diff
@@ -527,6 +542,17 @@ Rlist *RlistAppend(Rlist **start, const void *item, RvalType type)
         break;
 
     case RVAL_TYPE_LIST:
+        if (allow_all_types)
+        {
+            JsonElement* store = JsonArrayCreate(RlistLen(item));
+            for (const Rlist *rp = item; rp; rp = rp->next)
+            {
+                JsonArrayAppendElement(store, RvalToJson(rp->val));
+            }
+
+            return RlistAppendRval(start, (Rval) { store, RVAL_TYPE_CONTAINER });
+        }
+
```

Comment:
I'm a little confused by this one: If I understand correctly, the old code (`!allow_all_types`) appends element by element, whereas this appends only one element, a JSON structure. So `!allow_all_types` with two lists, `[1, 2, 3]` and `[4, 5, 6]`, would give `[1, 2, 3, 4, 5, 6]`, whereas `allow_all_types` would give `[1, 2, 3, [4, 5, 6]]`. Is this correct? If so, it's a little unexpected I think. Can you provide a bit more background for this change?